apache-http-client/src/main/java/com/joyent/http/signature/apache/httpclient/HttpSignatureCache.java

```diff
@@ -0,0 +1,134 @@
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Class that an interface to caching HTTP signatures for a given
+ * {@link Credentials} instance.
+ *
+ * @author <a href="https://github.com/dekobon">Elijah Zupancic</a>
+ * @since 3.0.0
+ */
+class HttpSignatureCache {
+    /**
+     * Credentials to associate cache with.
+     */
+    private final Credentials credentials;
+
+    /**
+     * Reference to the date time of the last HTTP signature.
+     */
+    private final AtomicReference<String> lastDate = new AtomicReference<>("");
+
+    /**
+     * Reference to the last generated HTTP signature associated with this credential.
+     */
+    private final AtomicReference<String> lastSignature = new AtomicReference<>("");
+
+    /**
+     * Creates a new cache for the specified credential.
+     *
+     * @param credentials credentials to associate cache with
+     */
+    HttpSignatureCache(final Credentials credentials) {
+        if (credentials == null) {
+            throw new IllegalArgumentException("Credentials must be present");
+        }
+
+        if (credentials.getPassword() == null) {
+            throw new IllegalArgumentException("Password (fingerprint) must be present");
+        }
+
+        if (credentials.getUserPrincipal() == null) {
+            throw new IllegalArgumentException("User principal must be present");
+        }
+
+        if (credentials.getUserPrincipal().getName() == null) {
+            throw new IllegalArgumentException("User principal name must be present");
+        }
+
+        this.credentials = credentials;
+    }
+
+    /**
+     * Method that attempts to get a valid cached signature based on the
+     * specified parameters. If a cached credential is not valid, then
+     * a new signature will be generated and stored for later use in caching.
+     *
+     * @param stringDate date to use for HTTP signature
+     * @param signer signer class to use for generating signature
+     * @param keyPair cryptographic key pair to use for generating signature
+     *
+     * @return a valid HTTP signature string to be passed as a header value
+     *
+     * @throws AuthenticationException thrown if there is a problem authenticating the signature
+     */
+    String updateAndGetSignature(final String stringDate,
+                                 final Signer signer,
+                                 final KeyPair keyPair) throws AuthenticationException {
+
+        if (lastDate.get().equals(stringDate)) {
+            final String authz = lastSignature.get();
+            if (!authz.isEmpty()) {
+                return authz;
+            }
+            // if signature is empty, we fall below and store the new value
+        }
+
+        synchronized (this) {
+            lastDate.set(stringDate);
+
+            final String login = credentials.getUserPrincipal().getName();
+            final String fingerprint = credentials.getPassword();
```

Comment:
I *think* this is just simple field access to keep things within the line length (so it is fine), but formally they don't need to be in the critical section.